Change setNumThreads to wait for thread start

kdt3rd opened the pull request:
Fix Issue 890, issue with windows shutdown when exiting quickly prior to threads actually starting. This section of setNumThreads will only run when actually changing the count, which limits the impact

Based on a proposed patch by Dieter De Baets @debaetsd

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Change setNumThreads to wait for thread start

Fix Issue 890, issue with windows shutdown when exiting quickly prior to
threads actually starting. This section of setNumThreads will only run
when actually changing the count, which limits the impact

Based on a proposed patch by Dieter De Baets @debaetsd

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

meshula reviewed (approved):
this change looks easier to reason about than the existing code.

kdt3rd committed:
Simplify logic in default threadpool provider

- avoid potential issue if one thread uses setNumThreads while another
calls setProvider
- remove unneeded sync object
- ensure that all threads are signalled when they need to shutdown

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Restore wait for threads to start

After simplifying, win32 and vp19 linux still crashes. Restore a wait
for the thread to start

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Add additional pointer checks and cleanup

Simplify the logic in setNumThreads, add pointer checks and disable the
wait for thread start, hunting for why win32 crashes on exit randomly

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

meshula commented:
I would like to land this, but note that 8 tests are failing. Do we know what's going on there?

kdt3rd committed:
simplify addTask in default thread provider given threadpool behavior

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Add null guard check to avoid crash in api function

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Further simplify launch / finish of threaded default worker pool, clean up legacy workarounds

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Further simplify default thread provider to avoid member access hades

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
fix missed data member in win32 block

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Fix test for older linux variant, disable windows workaround

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
The async startup issues seems to be wider spread than anticipated, enable globally for now

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Clean out ifdefs, update comments, restore wait at setNumThreads

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd commented:
> I would like to land this, but note that 8 tests are failing. Do we know what's going on there?

as I mentioned in the other issue, I do not have a good explanation for why I see some (consistent) crashes in some of the configs - what I've done shouldn't cause any. And now github has stopped updating this PR and rebuilding (I've pushed a commit to my tree and it won't start a CI run). Definitely stopping for the night now, will pick up tomorrow

kdt3rd committed:
Remove old code, replace with atomic_load

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
remove using std namespace, cache default provider ptr

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Fix compile when threading disabled

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Add experiment to always reset thread lists

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Revert experiment as not working

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Add LD_PRELOAD of libSegFault.so to tests to see where segfaults are happening

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
revert LD_PRELOAD check, apparently doesn't work in github containers

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Divide and conquer (problem not recreatable in local docker container)

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Divide and conquer 2

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Divide and Conquer 3: Pass data block to child thread to prevent crash

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

meshula commented:
This one's interesting, in the mac run


```
103/122 Test #103: OpenEXR.testTiledLineOrder ................***Exception: SegFault 31.09 sec
[21724](https://github.com/AcademySoftwareFoundation/openexr/actions/runs/4751940657/jobs/8441682047?pr=1291#step:7:21725)
tempDir = /var/tmp/OpenEXRTest_EHKKQBJS
```

kdt3rd committed:
Divide and conquer 4: change to only grow thread count

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd commented:
> This one's interesting, in the mac run
> 
> ```
> 103/122 Test #103: OpenEXR.testTiledLineOrder ................***Exception: SegFault 31.09 sec
> [21724](https://github.com/AcademySoftwareFoundation/openexr/actions/runs/4751940657/jobs/8441682047?pr=1291#step:7:21725)
> tempDir = /var/tmp/OpenEXRTest_EHKKQBJS
> ```

yeah, all of them are crashing in one way or another, except I can't recreate that, can't find anything with valgrind / helgrind, with any of the sanitizers. Rewrite the threading code entirely, same crashes. Disable threading, it's ok :-/

meshula commented:
> Rewrite the threading code entirely, same crashes

oof!!!

kdt3rd committed:
re-do finish and how task group finish wait happens

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Cleanup after experimentation

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Fix speling

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd commented:
> oof!!!

So the changes I made were fine. However, I removed enough overhead at the same time, I uncovered some fun un-defined behavior in TaskGroup (which I hadn't changed). Funny enough, it was sort of mentioned in reference to what was blamed as an old glibc bug previously, with a check to add another lock primitive. However, it was more insidious than that, and was affecting all platforms. I have changed TaskGroup to address this without doing the extra lock primitive such that we do it all the time now. As a bonus, the ThreadPool code is now a lot simpler than it was, even in my first attempt in this PR.

Assuming my latest cleanup also passes again, I will merge this tomorrow - will need to do a squash to cleanup all the "is this it" commits where I was in discovery mode. need to make a tool to load up my cpus to run tests with...
